### ISSUE TITLE
Customized loss function requires eager tensor, but symbolic tensor is passed

### ISSUE THREAD

**Issue by chuanli11** [tensorflow/tensorflow#39702]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): tensorflow-gpu==2.2.0
- Python version:3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1/7.6.5
- GPU model and memory: Quadro RTX 8000 / 48GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"` v2.2.0-rc4-8-g2b96f3662b 2.2.0

**Describe the current behavior**
I need to apply a binary mask to the model output for computing loss. My current implementation uses a model that takes two inputs (the data and the mask), and use function closure to implement the customized loss. 

However, this raises the error "tensorflow.python.eager.core._SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors".

Apparently, the mask input is treated as symbolic tensor.

**Describe the expected behavior**

This only happens in the eager mode. Apply `disable_eager_execution()` will eliminate the problem. However, I want to know if there is any way to make this work in the eager mode.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

This is the [gist](https://colab.research.google.com/drive/1D_GitVtknoYk33hURTXzC0murA60UBH0?usp=sharing)

```
import tensorflow as tf
import numpy as np

import tensorflow.keras.backend as K

from tensorflow.keras.layers import Input, Flatten, Dense
from tensorflow.keras import Model

x_data = np.zeros((32, 28, 28))
x_mask = np.zeros((32, 10))
y = np.zeros((32, 10))

input_data = Input(shape=(28, 28))
input_mask = Input(shape=(10,))

output= Flatten()(input_data)
output = Dense(64, activation='relu')(output)
output = Dense(10)(output)
model = Model(inputs=[input_data, input_mask], outputs=output)


def custom_loss():
    def loss(y_true, y_pred):
        # This line causes the error
        return K.mean(K.square(y_true - y_pred * model.inputs[1]), axis=-1)   

        # This line doesn't cause the error
        # return K.mean(K.square(y_true - y_pred), axis=-1)   
    return loss


model.compile(
    optimizer=tf.keras.optimizers.SGD(),
    loss=custom_loss(),
    metrics=['accuracy'])

for i in range (2):
    print(i)
    model.train_on_batch([x_data, x_mask], y)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.2 ,nightly version(2.3.0-dev20200519) and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/fb1237bf883632c27c2c617d100fbf44/untitled913.ipynb).Thanks!

**Comment by omalleyt12** [comment]
@chuanli11 Thanks for the issue!

Yes, the `model.inputs` are symbolic Tensors. For Functional Models, these Tensors are used to *build* the Model with a static graph, but in eager mode the Model is then *executed* with a `tf.function`. This means that `model.inputs` can't be used in losses, metrics, etc.

Instead, here's how I'd recommend achieving your use case. Essentially, `sample_weight` will handle this, rather than trying to access the mask an a `keras.Input`:

```python
import tensorflow as tf
import numpy as np

from tensorflow.keras.layers import Input, Flatten, Dense
from tensorflow.keras import Model

x_data = np.zeros((32, 28, 28))
x_mask = np.zeros((32, 10))
y = np.zeros((32, 10))

input_data = Input(shape=(28, 28))
output= Flatten()(input_data)
output = Dense(64, activation='relu')(output)
output = Dense(10)(output)
model = Model(inputs=input_data, outputs=output)


class MyLoss(tf.keras.losses.Loss):
  def call(self, y_true, y_pred):
      return (y_true - y_pred) ** 2  


model.compile(
    optimizer=tf.keras.optimizers.SGD(),
    loss=MyLoss(name='loss'),
    metrics=['accuracy'])

for i in range (2):
    print(i)
    loss = model.train_on_batch(x_data, y, sample_weight=x_mask)
```

Hope that helps!

Closing out as this is intended behavior

**Comment by chuanli11** [comment]
@omalleyt12 Thank you. This is really helpful.

**Comment by lastmansleeping** [comment]
@omalleyt12 I'm in a similar situation but I can not convert my code to not use model.inputs. Is there any other workaround for this?

**Comment by lastmansleeping** [comment]
Tagging @amahendrakar and @jvishnuvardhan as you had tagged this solution in other issue threads.

**Comment by jvishnuvardhan** [comment]
@lastmansleeping Can you please create a new issue with the standalone code you provided above? Thanks!
